# Python for Data Analysis `Lesson03`

#### Tags
* Título: _Case Study 02_
* Autor: AH Uyekita
* Data: 10/12/2018
* Cod: ND110

## Pratice

In this Lesson I hope I can have a bit more pratice working with the libraries, such as:

* Numpy
* Pandas
* Matplotlib

The last Lesson was very good to introduce them.

## Archives

All archives area downloaded from the EPA website.

* https://www.fueleconomy.gov/feg/download.shtml
* Table: Downloadable Fuel Economy Data
    * Column: EPA Green Vehicle Guide
    * File: TXT

### Alocating

These two dataset are located in the 01-Dataset folder.

* ~/01-Dataset/all_alpha_08.txt
* ~/01-Dataset/all_alpha_18.txt

## Reading the Guide Datafile

This file could be accessed/visited in:

* https://www.fueleconomy.gov/feg/EPAGreenGuide/GreenVehicleGuideDocumentation.pdf

Is also stored in ~/01-Dataset folder.

### Variables Description

In this file I could read a general information about each variable.

## Python Requeriments

Packages used to perform this study.

In [1]:
import pandas as pd
import numpy as np
import matplotlib as plt
import seaborn as sns
% matplotlib inline

## Loading Data

Creating two dataframe to store these two archives.

In [2]:
# Loading the EPA Fuel Economy Data from 2008.
car_2008 = pd.read_csv('01-Dataset/all_alpha_08.txt', sep="\t") # Atention to the sep, in this case is a tabulation.

# Loading the EPA Fuel Economy Data from 2018.
car_2018 = pd.read_csv('01-Dataset/all_alpha_18.txt', sep="\t") # Atention to the sep, in this case is a tabulation.

## Assessing Data

I will make some subsettings and minor analysis, just to investigate the structure of the dataset.

### Data Structure

#### First and last rows

In [3]:
car_2008.head(5)

Model  Displ      Cyl    Trans Drive      Fuel Sales Area Stnd  \
0  ACURA MDX    3.7  (6 cyl)  Auto-S5   4WD  Gasoline         CA   U2   
1  ACURA MDX    3.7  (6 cyl)  Auto-S5   4WD  Gasoline         FA   B5   
2  ACURA RDX    2.3  (4 cyl)  Auto-S5   4WD  Gasoline         CA   U2   
3  ACURA RDX    2.3  (4 cyl)  Auto-S5   4WD  Gasoline         FA   B5   
4   ACURA RL    3.5  (6 cyl)  Auto-S5   4WD  Gasoline         CA   U2   

   Underhood ID    Veh Class Air Pollution Score FE Calc Appr City MPG  \
0  8HNXT03.7PKR          SUV                   7          Drv       15   
1  8HNXT03.7PKR          SUV                   6          Drv       15   
2  8HNXT02.3DKR          SUV                   7          Drv       17   
3  8HNXT02.3DKR          SUV                   6          Drv       17   
4  8HNXV03.5HKR  midsize car                   7          Drv       16   

  Hwy MPG Cmb MPG  Unadj Cmb MPG Greenhouse Gas Score SmartWay  
0      20      17        22.0527                    4       no  
1      20      17        22.0527                    4       no  
2      22      19        24.1745                    5       no  
3      22      19        24.1745                    5       no  
4      24      19        24.5629                    5       no

In [4]:
car_2018.head(5)

Model  Displ  Cyl       Trans Drive      Fuel Cert Region       Stnd  \
0  ACURA ILX    2.4  4.0       AMS-8   2WD  Gasoline          CA  L3ULEV125   
1  ACURA ILX    2.4  4.0       AMS-8   2WD  Gasoline          FA     T3B125   
2  ACURA MDX    3.0  6.0       AMS-7   4WD  Gasoline          CA  L3ULEV125   
3  ACURA MDX    3.0  6.0       AMS-7   4WD  Gasoline          FA     T3B125   
4  ACURA MDX    3.5  6.0  SemiAuto-9   2WD  Gasoline          CA  L3ULEV125   

             Stnd Description  Underhood ID  Veh Class  Air Pollution Score  \
0  California LEV-III ULEV125  JHNXV02.4NH3  small car                    3   
1      Federal Tier 3 Bin 125  JHNXV02.4NH3  small car                    3   
2  California LEV-III ULEV125  JHNXV03.08H3  small SUV                    3   
3      Federal Tier 3 Bin 125  JHNXV03.08H3  small SUV                    3   
4  California LEV-III ULEV125  JHNXV03.5RH3  small SUV                    3   

  City MPG Hwy MPG Cmb MPG  Greenhouse Gas Score SmartWay Comb CO2  
0       25      35      29                     6       No      309  
1       25      35      29                     6       No      309  
2       26      27      27                     6       No      330  
3       26      27      27                     6       No      330  
4       19      27      22                     4       No      404

In [5]:
car_2008.tail(5)

Model  Displ      Cyl    Trans Drive      Fuel Sales Area Stnd  \
2399  VOLVO XC 90    3.2  (6 cyl)  Auto-S6   4WD  Gasoline         CA   U2   
2400  VOLVO XC 90    4.4  (8 cyl)  Auto-S6   4WD  Gasoline         CA   U2   
2401  VOLVO XC 90    3.2  (6 cyl)  Auto-S6   2WD  Gasoline         FA   B5   
2402  VOLVO XC 90    3.2  (6 cyl)  Auto-S6   4WD  Gasoline         FA   B5   
2403  VOLVO XC 90    4.4  (8 cyl)  Auto-S6   4WD  Gasoline         FA   B5   

      Underhood ID Veh Class Air Pollution Score FE Calc Appr City MPG  \
2399  8VVXB03.2U2N       SUV                   7          Drv       14   
2400  8VVXB04.4U2N       SUV                   7          Drv       13   
2401  8VVXB03.2U2N       SUV                   6          Drv       14   
2402  8VVXB03.2U2N       SUV                   6          Drv       14   
2403  8VVXB04.4U2N       SUV                   6          Drv       13   

     Hwy MPG Cmb MPG  Unadj Cmb MPG Greenhouse Gas Score SmartWay  
2399      20      16        20.7134                    4       no  
2400      19      15        19.4789                    3       no  
2401      20      16        20.9644                    4       no  
2402      20      16        20.7134                    4       no  
2403      19      15        19.4789                    3       no

In [6]:
car_2018.tail(5)

Model  Displ  Cyl       Trans Drive                  Fuel  \
2677  VOLVO XC 90    2.0  4.0  SemiAuto-8   4WD              Gasoline   
2678  VOLVO XC 90    2.0  4.0  SemiAuto-8   4WD              Gasoline   
2679  VOLVO XC 90    2.0  4.0  SemiAuto-8   4WD              Gasoline   
2680  VOLVO XC 90    2.0  4.0  SemiAuto-8   4WD  Gasoline/Electricity   
2681  VOLVO XC 90    2.0  4.0  SemiAuto-8   4WD  Gasoline/Electricity   

     Cert Region       Stnd            Stnd Description  Underhood ID  \
2677          FA      T3B70       Federal Tier 3 Bin 70  JVVXJ02.0A70   
2678          CA   L3ULEV70   California LEV-III ULEV70  JVVXJ02.0B70   
2679          FA      T3B70       Federal Tier 3 Bin 70  JVVXJ02.0B70   
2680          CA  L3SULEV30  California LEV-III SULEV30  JVVXJ02.0P30   
2681          FA      T3B30       Federal Tier 3 Bin 30  JVVXJ02.0P30   

         Veh Class  Air Pollution Score City MPG Hwy MPG Cmb MPG  \
2677  standard SUV                    5       22      28      24   
2678  standard SUV                    5       20      27      23   
2679  standard SUV                    5       20      27      23   
2680  standard SUV                    7    26/63   30/61   27/62   
2681  standard SUV                    7    26/63   30/61   27/62   

      Greenhouse Gas Score SmartWay Comb CO2  
2677                     5       No      365  
2678                     5       No      394  
2679                     5       No      394  
2680                    10    Elite      187  
2681                    10    Elite      187

### Variables

Let's plot the dimensions of each dataframe:

* Number of rows (observations);
* Number of columns (variables or features).

In [7]:
# Dimensions of the 2008 dataframe
dim_2008 = car_2008.shape

# Dimensions of the 2018 dataframe
dim_2018 = car_2018.shape

print("* 2008\nRows: {}\nColumns: {}\n\n* 2018\nRows: {}\nColumns: {}".format(*dim_2008,*dim_2018)) # Atention to the unpacking the tuples

* 2008
Rows: 2404
Columns: 18

* 2018
Rows: 2682
Columns: 18


From 2008 to 2018 the number of observations (it means: number of car models) has increased. Curiously, the number of variable kept the same.

Let's analyze the variables (or features) of these dataframes.

### Variables (Columns names)

I will bind these two dataframe into one, so I need to check if the columns has the same names, to be able to do bind them.

In [8]:
sorted(list(car_2008.columns))

['Air Pollution Score',
 'City MPG',
 'Cmb MPG',
 'Cyl',
 'Displ',
 'Drive',
 'FE Calc Appr',
 'Fuel',
 'Greenhouse Gas Score',
 'Hwy MPG',
 'Model',
 'Sales Area',
 'SmartWay',
 'Stnd',
 'Trans',
 'Unadj Cmb MPG',
 'Underhood ID',
 'Veh Class']

In [9]:
sorted(list(car_2018.columns))

['Air Pollution Score',
 'Cert Region',
 'City MPG',
 'Cmb MPG',
 'Comb CO2',
 'Cyl',
 'Displ',
 'Drive',
 'Fuel',
 'Greenhouse Gas Score',
 'Hwy MPG',
 'Model',
 'SmartWay',
 'Stnd',
 'Stnd Description',
 'Trans',
 'Underhood ID',
 'Veh Class']

Although, the absolute column number is equal, there are some new variables and dropped variables.

* New
    * Cert Region
    * Comb CO2
    * Stnd Description
* Dropped
    * FE Calc Appr
    * Sales Area
    * Unadj Cmb MPG

In the Parte 03 Lesson03 Exercise 4, I realized that `Cert Region` and `Sales Area` are the same thing. I will show the table with a very good summary of each variable (This is a copy and paste of the class notes).

# COPY and PASTE table

Following the good pratices, I will modify these columns names to lowercase and all space will be changed to underscores.

In [10]:
# The rename methods deals with changin spaces to underscores and lower casing everything.
car_2008.rename(columns=lambda x: x.strip().lower().replace(" ", "_"), inplace = True)

# The rename methods deals with changin spaces to underscores and lower casing everything.
car_2018.rename(columns=lambda x: x.strip().lower().replace(" ", "_"), inplace = True)

### NA values

Using the `.info()` I can see the number of non-full elements.

In [11]:
car_2008.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2404 entries, 0 to 2403
Data columns (total 18 columns):
model                   2404 non-null object
displ                   2404 non-null float64
cyl                     2205 non-null object
trans                   2205 non-null object
drive                   2311 non-null object
fuel                    2404 non-null object
sales_area              2404 non-null object
stnd                    2404 non-null object
underhood_id            2404 non-null object
veh_class               2404 non-null object
air_pollution_score     2404 non-null object
fe_calc_appr            2205 non-null object
city_mpg                2205 non-null object
hwy_mpg                 2205 non-null object
cmb_mpg                 2205 non-null object
unadj_cmb_mpg           2205 non-null float64
greenhouse_gas_score    2205 non-null object
smartway                2404 non-null object
dtypes: float64(2), object(16)
memory usage: 338.1+ KB


In [12]:
car_2018.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2682 entries, 0 to 2681
Data columns (total 18 columns):
model                   2682 non-null object
displ                   2637 non-null float64
cyl                     2637 non-null float64
trans                   2682 non-null object
drive                   2682 non-null object
fuel                    2682 non-null object
cert_region             2682 non-null object
stnd                    2682 non-null object
stnd_description        2682 non-null object
underhood_id            2682 non-null object
veh_class               2682 non-null object
air_pollution_score     2682 non-null int64
city_mpg                2682 non-null object
hwy_mpg                 2682 non-null object
cmb_mpg                 2682 non-null object
greenhouse_gas_score    2682 non-null int64
smartway                2682 non-null object
comb_co2                2682 non-null object
dtypes: float64(2), int64(2), object(14)
memory usage: 377.2+ KB


Based on these two `.info()` results, I can highlight:

* Number of entries in 2008: 2404
    * In some cases, variables has less (it means: NA cells)
* Number of entries in 2018: 2682
    * In some cases, variables has less (it means: NA cells)

So, these two dataframe has `NA` values.

* 2008: 199 NA rows
* 2018: 45 NA rows

### Duplicated entries

I would like to know if there are any duplicated rows.

In [13]:
# Calculation of duplicated rows in each dataframe.
dupe_2008, dupe_2018 = sum(car_2008.duplicated()),sum(car_2018.duplicated())

print("Duplicated rows in 2008: {}\nDuplicated rows in 2008: {}".format(dupe_2008, dupe_2018))

Duplicated rows in 2008: 25
Duplicated rows in 2008: 0


As you can see, there are 25 rows duplicated in 2008.

### Summarization

A very short way to summarize the variable in each dataframe, is using the method `.describre()`.

In [14]:
car_2008.describe()

displ  unadj_cmb_mpg
count  2404.000000    2205.000000
mean      3.748918      23.916104
std       1.335785       6.366170
min       1.300000      10.018400
25%       2.500000      19.113900
50%       3.500000      23.921300
75%       4.800000      27.869300
max       8.400000      65.777800

In [15]:
car_2018.describe()

displ          cyl  air_pollution_score  greenhouse_gas_score
count  2637.000000  2637.000000          2682.000000           2682.000000
mean      3.017482     5.495639             4.133855              4.834452
std       1.303469     1.827874             1.959773              1.895488
min       0.600000     2.000000             1.000000              1.000000
25%       2.000000     4.000000             3.000000              4.000000
50%       2.900000     6.000000             3.000000              5.000000
75%       3.600000     6.000000             5.000000              6.000000
max       8.000000    16.000000            10.000000             10.000000

The results are quite distincts, due to the variable class type, this summary only works to `int` or `float` variables.

### Unique attributes

Sometimes It is useful to analyze the number of unique (this is the reason I am using `.nunique()`) values in each variable. In case of categorical variable, I can see if a new category is "released".

In [16]:
# Number of Unique values
car_2008.nunique()

model                   436
displ                    47
cyl                       8
trans                    14
drive                     2
fuel                      5
sales_area                3
stnd                     12
underhood_id            343
veh_class                 9
air_pollution_score      13
fe_calc_appr              2
city_mpg                 39
hwy_mpg                  43
cmb_mpg                  38
unadj_cmb_mpg           721
greenhouse_gas_score     20
smartway                  2
dtype: int64

In [17]:
# Number of Unique values
car_2018.nunique()

model                   645
displ                    45
cyl                       9
trans                    29
drive                     2
fuel                      6
cert_region               2
stnd                     25
stnd_description         25
underhood_id            418
veh_class                10
air_pollution_score       7
city_mpg                113
hwy_mpg                 118
cmb_mpg                 109
greenhouse_gas_score     10
smartway                  3
comb_co2                405
dtype: int64

An example to see the difference between these two dataframes is analyzing the `fuel` variable. Now, I will use `.unique()`.

In [18]:
# Number of Unique values
car_2008['fuel'].unique()

array(['Gasoline', 'ethanol/gas', 'ethanol', 'diesel', 'CNG'],
      dtype=object)

In [19]:
# Number of Unique values
car_2018['fuel'].unique()

array(['Gasoline', 'Gasoline/Electricity', 'Diesel', 'Electricity',
       'Ethanol/Gas', 'Hydrogen'], dtype=object)

There are new categories and other were dropped. Clearly, I must lowercase and remove the spaces to follow a better standard of coding.

## Cleaning Column Labels

### Q1: Drop extraneous columns


Just to make sure all the names are equal, and removing unnecessary variables.

Columns to Drop:

* From 2008 dataset
    * 'Stnd', 'Underhood ID', 'FE Calc Appr', 'Unadj Cmb MPG'
* From 2018 dataset
    * 'Stnd', 'Stnd Description', 'Underhood ID', 'Comb CO2'

In [20]:
# Removing unnecessary columns, so dropping columns from 2008 dataset
car_2008.drop(['stnd', 'underhood_id', 'fe_calc_appr', 'unadj_cmb_mpg'], axis=1, inplace=True) # Will update the current dataframe

In [21]:
# Removing unnecessary columns, so dropping columns from 2018 dataset
car_2018.drop(['stnd', 'stnd_description', 'underhood_id', 'comb_co2'], axis=1, inplace=True) # Will update the current dataframe

Now, both dataframe has the same variables with same names.

### Q2: Rename Columns

* Change the "Sales Area" column label in the 2008 dataset to "Cert Region" for consistency.
* Rename all column labels to replace spaces with underscores and convert everything to lowercase. (Underscores can be much easier to work with in Python than spaces. For example, having spaces wouldn't allow you to use df.column_name instead of df['column_name'] to select columns or use query(). Being consistent with lowercase and underscores also helps make column names easy to remember.)

In this step I need to lowercase and remove the spaces in the columns names, but I have anticepated it. I only need to rename the `Sales Region` to `Cert Region`.

In [22]:
# rename Sales Area to Cert Region
names_columns_2008 = list(car_2008.columns)  # Save name columns
names_columns_2008[6] = 'cert_region'        # Update Sales Area to Cert Region

# Renaming columns names, now with Cert Region.
car_2008.columns = names_columns_2008

# confirm changes
car_2008.head(1)

model  displ      cyl    trans drive      fuel cert_region veh_class  \
0  ACURA MDX    3.7  (6 cyl)  Auto-S5   4WD  Gasoline          CA       SUV   

  air_pollution_score city_mpg hwy_mpg cmb_mpg greenhouse_gas_score smartway  
0                   7       15      20      17                    4       no

The columns are equal.

In [23]:
# confirm column labels for 2008 and 2018 datasets are identical
car_2008.columns == car_2018.columns

array([ True,  True,  True,  True,  True,  True,  True,  True,  True,
        True,  True,  True,  True,  True])

Just to make sure everything is equal!

In [24]:
# make sure they're all identical like this
(car_2008.columns == car_2018.columns).all()

True

## Filter, Drop Nulls, Dedupe

1. Filter
For consistency, only compare cars certified by California standards. Filter both datasets using `query` to select only rows where `cert_region` is `CA`. Then, drop the `cert_region` columns, since it will no longer provide any useful information (we'll know every value is 'CA').

2. Drop Nulls
Drop any rows in both datasets that contain missing values.

3. Dedupe
Drop any duplicate rows in both datasets.

In [40]:
car_2008.shape

(2404, 14)

In [41]:
car_2018.shape

(2682, 14)

### Q1: Filter by Certification Region

I will subset the original dataset to a small one, with only California data.

In [25]:
# filter datasets for rows following California standards
df_08 = car_2008.query('cert_region == "CA"')
df_18 = car_2018.query('cert_region == "CA"')

Keep in mind that:

```
my_filter = "CA"
car_2008.query('cert_region == "{}"'.format(my_filter))
```

Also works here.

Just to be sure the filter work, let's find the uniques values of `cert_region`.

In [26]:
# confirm only certification region is California
df_08['cert_region'].unique()

array(['CA'], dtype=object)

In [27]:
# confirm only certification region is California
df_18['cert_region'].unique()

array(['CA'], dtype=object)

Dropping the variable `cert_region`, because it is not a useful variable, due to has only `CA` values.

In [28]:
# drop certification region columns form both datasets
df_08 = df_08.drop('cert_region', axis=1) 

In [29]:
df_18 = df_18.drop('cert_region', axis=1)

Checking the dimensions of this dataframe after drop one variable.

In [30]:
df_08.shape

(1084, 13)

In [31]:
df_18.shape

(1322, 13)

### Q2: Drop Rows with Missing Values

Let's shows the number of entries and non-full values.

In [32]:
# view missing value count for each feature in 2008
df_08.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1084 entries, 0 to 2400
Data columns (total 13 columns):
model                   1084 non-null object
displ                   1084 non-null float64
cyl                     1009 non-null object
trans                   1009 non-null object
drive                   1047 non-null object
fuel                    1084 non-null object
veh_class               1084 non-null object
air_pollution_score     1084 non-null object
city_mpg                1009 non-null object
hwy_mpg                 1009 non-null object
cmb_mpg                 1009 non-null object
greenhouse_gas_score    1009 non-null object
smartway                1084 non-null object
dtypes: float64(1), object(12)
memory usage: 118.6+ KB


In [33]:
# view missing value count for each feature in 2018
df_18.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1322 entries, 0 to 2680
Data columns (total 13 columns):
model                   1322 non-null object
displ                   1302 non-null float64
cyl                     1302 non-null float64
trans                   1322 non-null object
drive                   1322 non-null object
fuel                    1322 non-null object
veh_class               1322 non-null object
air_pollution_score     1322 non-null int64
city_mpg                1322 non-null object
hwy_mpg                 1322 non-null object
cmb_mpg                 1322 non-null object
greenhouse_gas_score    1322 non-null int64
smartway                1322 non-null object
dtypes: float64(2), int64(2), object(9)
memory usage: 144.6+ KB


There are some `NA` values, by using the `.dropna()` method I can remove these lines.

In [34]:
# drop rows with any null values in both datasets
df_08.dropna(axis=0, inplace = True)
df_18.dropna(axis=0, inplace = True)

The method `.isnull()` checks if some row has null values.

In [35]:
# checks if any of columns in 2008 have null values - should print False
df_08.isnull().sum().any()

False

In [36]:
# checks if any of columns in 2018 have null values - should print False
df_18.isnull().sum().any()

False

### Q3: Dedupe Data

The method `.drop_duplicates()` will remove any row with at least one null value. First, let's print the number of duplicates entries in both dataframes.

In [37]:
# print number of duplicates in 2008 and 2018 datasets
dupe_2008, dupe_2018 = sum(df_08.duplicated()),sum(df_18.duplicated())

# Print
print("Duplicated in 2008: {}\nDuplicated in 2018: {}".format(dupe_2008, dupe_2018))

Duplicated in 2008: 23
Duplicated in 2018: 10


In [38]:
# drop duplicates in both datasets
df_08.drop_duplicates(inplace= True)
df_18.drop_duplicates(inplace= True)

Just to ensure if I have removed all the null rows.

In [42]:
# print number of duplicates again to confirm dedupe - should both be 0
dupe_2008, dupe_2018 = sum(df_08.duplicated()),sum(df_18.duplicated())

# Print
print("Duplicated in 2008: {}\nDuplicated in 2018: {}".format(dupe_2008, dupe_2018))

Duplicated in 2008: 0
Duplicated in 2018: 0


## Inspecting Data Types

Use this Jupyter Notebook to inspect the datatypes of features in each dataset and think about what changes should be made to make them practical and consistent (in both datasets). Use this to help you answer the quiz questions below.